File: Java/src/main/java/com/DSA/Java/Algorithms/Greedy/FractionalKnapsackProblem/FractionalKnapsack.java
package com.DSA.Java.Algorithms.Greedy.FractionalKnapsackProblem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class FractionalKnapsack {
  private ArrayList<KnapsackItem> items;

  public FractionalKnapsack() {
    this.items = new ArrayList<>();
  }

  public void addItem(int index, int value, int weight) {
    items.add(new KnapsackItem(index, value, weight));
  }

  public void knapSack(int capacity) {
    Comparator<KnapsackItem> comparator = new Comparator<KnapsackItem>() {
      @Override
      public int compare(KnapsackItem arg0, KnapsackItem arg1) {
        if (arg1.getRatio() > arg0.getRatio()) return 1;
        else return -1;
      }
    };

    Collections.sort(items, comparator);
    int usedCapacity = 0;
    double totalValue = 0;

    for (KnapsackItem item : items) {
      if (usedCapacity + item.getWeight() <= capacity) {
        usedCapacity += item.getWeight();
        System.out.println("Taken: " + item);
        totalValue += item.getValue();
      } else {
        int usedWeight = capacity - usedCapacity;
        double value = item.getRatio() * usedWeight;
        System.out.println("Taken: Item index = " + item.getIndex() + ", Obtained value = " + value +
            ", Used weight = " + usedWeight + ", Ratio = " + item.getRatio());
        usedCapacity += usedWeight;
        totalValue += value;
      }

      if (usedCapacity == capacity) break;
    }

    System.out.println("\nTotal value obtained : " + totalValue);
  }
}

File: C++/Algorithm/src/Greedy/FractionalKnapsackProblem/FractionalKnapsack.cpp
#include "../../../include/Greedy/FractionalKnapsackProblem/FractionalKnapsack.h"

void FractionalKnapsack::add_items(int index, int value, int weight) {
  items.push_back(new KnapsackItem(index, value, weight));
}

void FractionalKnapsack::knapsack(int capacity) {
  std::sort(items.begin(), items.end(),
            [] (KnapsackItem *item_1, KnapsackItem *item_2) {
              return item_1->get_ratio() > item_2->get_ratio();
            });

  int used_capacity {0};
  double total_value {0};

  std::cout << "\n";

  for (const auto &item : items) {
    if (used_capacity + item->get_weight() <= capacity) {
      used_capacity += item->get_weight();
      std::cout << "Taken : [ " << *item << " ]" << std::endl;
      total_value += item->get_value();
    } else {
      int used_weight = capacity - used_capacity;
      double value = item->get_ratio() * used_weight;
      std::cout << "Taken : [ " << *item << " ]" << ", Used weight : " << used_weight << std::endl;
      used_capacity += used_weight;
      total_value += value;
    }

    if (used_capacity == capacity) break;
  }

  std::cout << "\nTotal value obtained : " << total_value << std::endl;
}
